src/net_processing.cpp

```diff
@@ -921,11 +912,10 @@ bool AddOrphanTx(const CTransactionRef& tx, NodeId peer) EXCLUSIVE_LOCKS_REQUIRE
         return false;
     }
 
-    auto ret = mapOrphanTransactions.emplace(hash, COrphanTx{tx, peer, GetTime() + ORPHAN_TX_EXPIRE_TIME, g_orphan_list.size()});
+    auto ret = mapOrphanTransactions.emplace(hash, COrphanTx{tx, peer, GetTime() + ORPHAN_TX_EXPIRE_TIME});
     assert(ret.second);
-    g_orphan_list.push_back(ret.first);
     for (const CTxIn& txin : tx->vin) {
-        mapOrphanTransactionsByPrev[txin.prevout].insert(ret.first);
+        mapOrphanTransactionsByPrev[txin.prevout].insert(hash);
```

Comment:
> the lifetime of that object is only as long as the transaction exists in the mapOrphanTransactions

I think that is incorrect. The `hash` reference that we insert into `mapOrphanTransactionsByPrev` is created here:

https://github.com/bitcoin/bitcoin/blob/d02fcd0d45f428f73bd1de8a7a411fd34216d2fc/src/net_processing.cpp#L897

and the `GetHash()` method returns a reference to its `CTransaction::hash` member:

https://github.com/bitcoin/bitcoin/blob/d02fcd0d45f428f73bd1de8a7a411fd34216d2fc/src/primitives/transaction.h#L303

So, the reference is valid as long as the `CTransaction` object is valid (pointed to by the first argument of `AddOrphanTx()`).

Verifying whether the `CTransaction` object outlives the reference saved in `mapOrphanTransactionsByPrev` is a bit obscure.

Saving `uint256` instead of `uint256&` looks more robust, also considering possible future changes. It would require 32 bytes instead of 8 (+24 bytes per transaction).

Reply:
> the reference is valid as long as the CTransaction object is valid

Right, and I believe that for an orphan transaction, the only shared_ptr to the CTransaction is held in mapOrphanTransactions. As soon as it gets removed from that map, then the CTransaction will be destructed and the uint256& will be dangling.

> Saving uint256 instead of uint256& looks more robust, also considering possible future changes. It would require 32 bytes instead of 8 (+24 bytes per transaction).

This would require an additional 24 bytes per transaction input, not per transaction.

Reply:
> As soon as it gets removed from that map, then the CTransaction will be destructed and the uint256& will be dangling.

In `EraseOrphanTx()` a transaction is erased from `mapOrphanTransactions` only after its inputs have been erased from `mapOrphanTransactionsByPrev()`.

So, no dangling `uint256&` are expected.